index.js

```diff
@@ -75,23 +82,75 @@ function compression (options) {
 
     // proxy
 
-    res.write = function write (chunk, encoding) {
-      if (ended) {
+    res.write = function write (chunk, encoding, callback) {
+      if (chunk === null) {
+        // throw ERR_STREAM_NULL_VALUES
+        return _write.call(this, chunk, encoding, callback)
+      } else if (typeof chunk === 'string' || typeof chunk.fill === 'function' || isUint8Array(chunk)) {
+        // noop
+      } else {
+        // throw ERR_INVALID_ARG_TYPE
+        return _write.call(this, chunk, encoding, callback)
+      }
+
+      if (!callback && typeof encoding === 'function') {
+        callback = encoding
+        encoding = undefined
+      }
+
+      if (typeof callback !== 'function') {
+        callback = noop
+      }
+
+      if (res.destroyed || res.finished || ended) {
+        // HACK: node doesn't expose internal errors,
+        // we need to fake response to throw underlying errors type
+        var fakeRes = new ServerResponse({})
+        fakeRes.on('error', function (err) {
+          res.emit('error', err)
+        })
+        fakeRes.destroyed = res.destroyed
+        fakeRes.finished = res.finished || ended
+        // throw ERR_STREAM_DESTROYED or ERR_STREAM_WRITE_AFTER_END
+        _write.call(fakeRes, chunk, encoding, callback)
         return false
       }
 
       if (!this._header) {
         this._implicitHeader()
       }
 
+      if (chunk) {
+        chunk = toBuffer(chunk, encoding)
+        if (/^v0\.8\./.test(process.version) && stream) {
```

Comment:
Ideally we try to prefer feature testing rather than version testing. But at least this could be done once at the top of the lib and saved as a var instead of regexp testing every time.